### ISSUE TITLE
Evaluate if we can adapt our AoS Datastructures to SoA

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#5724]
https://www.intel.com/content/dam/www/public/us/en/documents/presentation/improving-vectorization-efficiency.pdf

**Comment by jpfeuffer** [comment]
Yes, when it gets to spectra it would also be much better for exporting to e.g. python..

**Comment by timosachsenberg** [comment]
Thanks for the analysis. My guess is that with SIMD enabled, many operations could benefit from this memory layout. I think @cbielow mentioned that we probably would need to add the compiler flag

**Comment by cbielow** [comment]
While I can see the benefits of SoA (like benefitting from SSE and maybe cache efficiency for searching), there are also downsides which should not be underestimated: sorting a SoA is much harder, i.e. requires manually keeping track of all arrays and using a helper structure to keep track of indices. Sorting is done all the time in OpenMS, and it will probably be less efficient and require more code. A similar argument applies for filtering/removing elements.

The benchmark above is not indicative of speed gains IMHO, because an MSSpectrum consists of much more than just the peak data (and thus calls a lot more C'tors and mallocs), whereas OSpectrum is just a bunch data arrays.
I would also suspect that, as long as the struct inside the array has members which are trivially copyable (such as double, float, int etc), the compiler will optimize the copy to be a `memmove`, which is indeed what happens: https://godbolt.org/z/McE5Esx9x
So in terms of copying, expect no speed gain when switching from AoS to SoA.

**Comment by jpfeuffer** [comment]
Sorting should be easy when implementing a zipview (or borrowing it from ranges-v3 library)

**Comment by timosachsenberg** [comment]
The question would be if we could also get rid of the SpectrumPtr implementation in OpenSWATH and unify the usage a bit more.
Potentially would mean that we would also need to change the way we currently store other meta data in the spectra as the OpenSWATH SpectraPtr are rather "thin" and don't provide all functionality as MSSpectrum.
